src/WebJobs.Extensions/Extensions/Timers/Scheduling/StorageScheduleMonitorV2.cs

```diff
@@ -0,0 +1,183 @@
+        /// Constructs a new instance.
+        /// </summary>
+        /// <param name="options"></param>
+        /// <param name="hostIdProvider"></param>
+        /// <param name="configuration"></param>
+        /// <param name="loggerFactory"></param>
+        public StorageScheduleMonitorV2(IOptions<JobHostInternalStorageOptions> options, IHostIdProvider hostIdProvider,
+            IConfiguration configuration, ILoggerFactory loggerFactory, AzureStorageProvider azureStorageProvider)
+        {
+            _storageOptions = options?.Value ?? throw new ArgumentNullException(nameof(options));
+            _configuration = configuration ?? throw new ArgumentNullException(nameof(configuration));
+            _hostIdProvider = hostIdProvider ?? throw new ArgumentNullException(nameof(hostIdProvider));
+            _logger = loggerFactory.CreateLogger(LogCategories.CreateTriggerCategory("Timer"));
+            _azureStorageProvider = azureStorageProvider ?? throw new ArgumentNullException(nameof(azureStorageProvider));
+
+            JsonSerializerSettings settings = new JsonSerializerSettings
+            {
+                DateFormatHandling = DateFormatHandling.IsoDateFormat
+            };
+            _serializer = JsonSerializer.Create(settings);
+        }
+
+        /// <summary>
+        /// Gets the blob directory where timer statuses will be stored.
+        /// </summary>
+        internal string TimerStatusPath
+        {
+            get
+            {
+                // We have to delay create the blob directory since we require the JobHost ID, and that will only
+                // be available AFTER the host as been started
+                if (string.IsNullOrEmpty(_timerStatusPath))
+                {
+                    string hostId = _hostIdProvider.GetHostIdAsync(CancellationToken.None).GetAwaiter().GetResult();
+                    if (string.IsNullOrEmpty(hostId))
+                    {
+                        throw new InvalidOperationException("Unable to determine host ID.");
+                    }
+
+                    _timerStatusPath = string.Format("timers/{0}", hostId);
+                }
+
+                return _timerStatusPath;
+            }
+        }
+
+        internal BlobContainerClient ContainerClient
+        {
+            get
+            {
+                if (_containerClient == null)
+                {
+                    if (_storageOptions.InternalSasBlobContainer != null)
+                    {
+                        _containerClient = new BlobContainerClient(new Uri(_storageOptions.InternalSasBlobContainer));   
+                    }
+                    else if (_storageOptions.InternalContainerName != null)
+                    {
+                        // TODO: Should we throw if we can't create the client?
+                        if (!_azureStorageProvider.TryGetBlobServiceClientFromConnection(out BlobServiceClient blobServiceClient, ConnectionStringNames.Storage))
```

Comment:
Should we make AzureStorageProvider aware of JobHostInternalStorageOptions?

Reply:
Hmm good point. It might be wise to consolidate all this logic with container overrides. @fabiocav to chime in.